pkg/action/list.go

```diff
@@ -150,6 +154,12 @@ func (l *List) Run() ([]*release.Release, error) {
 		return nil, err
 	}
 
+	if l.AllNamespaces {
+		if err := l.cfg.Init(settings.RESTClientGetter(), "", os.Getenv("HELM_DRIVER"), debug); err != nil {
```

Comment:
Regarding #11656 : 
There are two problems the issue raises. 
The first is that the namespace logic is in the helm/cli/list.go instead of pkg/action/list.go.
The second is that the solution depends on a mutating a configuration state.

The context:
Generally speaking; the code is used by CLI and the code could also be used as a package (hint: https://helm.sh/docs/topics/advanced/)

Both the master branch and this PR, do this by mutating the configuration.

For CLI: The code works because configuration is reset before each command.
https://github.com/helm/helm/blob/c137bfb68b4f8b7a10687f043b0dbccb09574969/cmd/helm/helm.go#L72-L75

The issue for package users:
The current implementation, puts requirements on the package users implementations. They _must_ reset the configuration each run.
If not: 
1. There is a risk of exposing unwanted namespaces due to either concurrent or subsequent uses of the client.
2. The code will break for subsequent use after AllNamespaces is utilized.

I don't think this is safe as it creates a trap with dangerous default behavior that gets triggered only when AllNamespaces is used, all other code with work without resetting the configuration.

Suggestion: 
I think the code should clone/create a new configuration that then is mutated as line 157.

Reply:
`if err := l.cfg.Init(settings.RESTClientGetter(), "", os.Getenv("HELM_DRIVER"), debug); err != nil {`
should be:
`if err := l.cfg.Init(cfg.RESTClientGetter(), "", os.Getenv("HELM_DRIVER"), debug); err != nil {`